### PR TITLE
Migrate repo to Typescript

### PR DESCRIPTION
**Pull request by jakecoppinger**
This PR migrates the repo to Typescript. It's a big change - if you'd rather not migrate I understand too.

The only issue I found after adding Typescript compilation to the Webpack config is a call to the imported JS libraries in `src/parking/interface.ts` must be done, otherwise the imported code is not run. I assume this is an optimisation to prevent code that isn't used from being called.

Please let me know if I could improve this :) I've added a lot of `: any` statements when I am unsure of the desired type. I may narrow down the type specificity in further PRs.

### CODE REVIEW

**Comment by jakecoppinger** [comment]
Ah I'm getting this issue in edit mode, currently looking into it:
```
Uncaught (in promise) TypeError: conditions.intervals is not iterable
at getColorByDate (parking-lane.ts:205)
```

**Comment by jakecoppinger** [comment]
Fixed edit bug where black line roads wouldn't appear - It seems TS conversion added some object prototypes in some objects for some reason. This changed the behaviour of the `!conditions` statement.

I've changed the statement to be more specific - checking if the object is empty.

File: src/utils/opening-hours.ts [hunk 1]
```diff
@@ -1,7 +1,8 @@
 import OpeningHours from 'opening_hours'
```
Comment (zlant) [inline src/utils/opening-hours.ts:1]: // @ts-ignore

File: src/parking/parking-lane.ts [hunk 2]
```diff
@@ -178,14 +203,20 @@ function getConditions(side, tags) {
...
+        const color = getColorByDate(lanes[lane].options.conditions, datetime)
+        lanes[lane].setStyle({ color })
+    }
 }
 
-function getColorByDate(conditions, datetime) {
-    if (!conditions)
+function getColorByDate(conditions: ConditionsInterface, datetime: Date): ConditionColor | undefined {
+    // Is object empty?
+    if (Object.keys(conditions).length === 0 || conditions.intervals === undefined) {
```
Comment (zlant) [inline src/parking/parking-lane.ts:217]: if (conditions == null) {

File: src/parking/parking-lane.ts [hunk 3]
```diff
@@ -178,14 +203,20 @@ function getConditions(side, tags) {
...
         return 'black'
+    }
 
     for (const interval of conditions.intervals) {
```
Comment (zlant) [inline src/parking/parking-lane.ts:221]: for (const interval of conditions.intervals ?? []) {

File: src/parking/parking-lane.ts (hunk 2 above)
Comment (zlant) [reply src/parking/parking-lane.ts:217]: lane 217

File: src/parking/parking-lane.ts (hunk 3 above)
Comment (zlant) [reply src/parking/parking-lane.ts:221]: lane 221

File: src/parking/parking-lane.ts (hunk 2 above)
Comment (zlant) [reply src/parking/parking-lane.ts:217]: If no intervals, then must be returned default condition color

**Comment by jakecoppinger** [comment]
I've:
- Bumped `opening_hours` from `^3.6.0` to `^3.7.0` to fix TS types
- Changed the logic so that when conditions.intervals is undefined, the default color is returned rather than `black`

Hope that's what you were looking for 😄

**Comment by jakecoppinger** [comment]
Thanks!

**Pull request merged**